src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -459,101 +473,38 @@ private void UnregisterRead()
+                Debug.Fail(string.Format("UnregisterRead failed: {0}", errorCode));
             }
         }
 
-        private void CloseInner()
+        public void Close()
         {
-            Debug.Assert(Monitor.IsEntered(_queueLock));
-
-            OperationQueue<AcceptOrConnectOperation> acceptOrConnectQueue;
-            OperationQueue<SendOperation> sendQueue;
-            OperationQueue<TransferOperation> receiveQueue;
-
-            // Drain queues and unregister events
-
-            acceptOrConnectQueue = _acceptOrConnectQueue.Stop();
-            sendQueue = _sendQueue.Stop();
-            receiveQueue = _receiveQueue.Stop();
-
-            Unregister();
-
-            // TODO: assert that queues are all empty if _registeredEvents was Interop.Sys.SocketEvents.None?
-
-            // TODO: the error codes on these operations may need to be changed to account for
-            //       the close. I think Winsock returns OperationAborted in the case that
-            //       the socket for an outstanding operation is closed.
-
-            Debug.Assert(!acceptOrConnectQueue.IsStopped || acceptOrConnectQueue.IsEmpty);
-            while (!acceptOrConnectQueue.IsEmpty)
+            lock (_queueLock)
             {
-                AcceptOrConnectOperation op = acceptOrConnectQueue.Head;
-                op.AbortAsync(this);
-                acceptOrConnectQueue.Dequeue();
-            }
+                // Drain queues
 
-            Debug.Assert(!sendQueue.IsStopped || sendQueue.IsEmpty);
-            while (!sendQueue.IsEmpty)
-            {
-                SendReceiveOperation op = sendQueue.Head;
-                op.AbortAsync(this);
-                sendQueue.Dequeue();
-            }
+                _acceptOrConnectQueue.StopAndAbort();
+                _sendQueue.StopAndAbort();
+                _receiveQueue.StopAndAbort();
 
-            Debug.Assert(!receiveQueue.IsStopped || receiveQueue.IsEmpty);
-            while (!receiveQueue.IsEmpty)
-            {
-                TransferOperation op = receiveQueue.Head;
-                op.AbortAsync(this);
-                receiveQueue.Dequeue();
-            }
-        }
+                // Freeing the token will prevent any future event delivery.  This socket will be unregistered
+                // from the event port automatically by the OS when it's closed.
+                _asyncEngineToken.Free();
 
-        public void Close()
-        {
-            lock (_queueLock)
-            {
-                CloseInner();
+                // TODO: assert that queues are all empty if _registeredEvents was Interop.Sys.SocketEvents.None?
+
+                // TODO: the error codes on these operations may need to be changed to account for
+                //       the close. I think Winsock returns OperationAborted in the case that
+                //       the socket for an outstanding operation is closed.
```

Comment:
I didn't add these TODOs, but I think I'll remove them - I'm not sure they're relevant any more.